### PR TITLE
refactor(module:core): add nz drag service

### LINKED ISSUE #4340: carousel pre and next not work
https://stackblitz.com/edit/ng-zorro-antd-start-gt9f39?file=src/app/app.component.ts

### PR DESCRIPTION
**Pull request by wendzhue**
close #4340

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Comment by wendzhue** [comment]
I will refactor other components in after-coming pull requests.

File: components/core/services/drag/nz-drag.service.ts [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+          x: e.pageX - this.currentStartingPoint!.x,
+          y: e.pageY - this.currentStartingPoint!.y
+        };
+      }),
+      filter((e: Delta) => Math.abs(e.x) > this.draggingThreshold || Math.abs(e.y) > this.draggingThreshold),
+      finalize(() => this.teardownDraggingSequence())
+    );
+  }
+
+  private registerDraggingHandler(): void {
```
Comment (hsuanxyz) [inline components/core/services/drag/nz-drag.service.ts]: 这里可以把事件带进来判断一下是监听 mouse 还是 touch.

File: components/core/services/drag/nz-drag.service.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    this.awaitingSequence = true;
+
+    // Complete last dragging sequence if a new target is dragged.
+    if (this.currentDraggingSequence) {
+      this.currentDraggingSequence.complete();
+    }
+
+    this.currentStartingPoint = getPagePosition(event);
+    this.currentDraggingSequence = new Subject<MouseEvent | Touch>();
+    this.awaitingSequence = false;
```
Comment (hsuanxyz) [inline components/core/services/drag/nz-drag.service.ts]: 这个变量应该没有用，从 `true` 到 `false` 中间全是同步的过程

Comment (wendzhue) [reply components/core/services/drag/nz-drag.service.ts]: 有用。

```ts
    if (this.currentDraggingSequence) {
      this.currentDraggingSequence.complete();
    }
```

这里 complete() 的时候会调用 `teardownDraggingSequence`。

不过这种防止多次绑定 handler 的机制倒是有点多余，用户连续两次拖拽不可能是同步事件。我这里给判断加一个延时好了。

**Pull request merged**